### PR TITLE
Cleaning up context usage in rop and srop

### PR DESCRIPTION
**Pull request by eQu1NoX**
Cleaning up context usage in rop and srop + functions internal to srop

### CODE REVIEW

File: pwnlib/srop.py [hunk 1]
```diff
@@ -1,26 +1,27 @@
 from .util.packing import pack
+from .context import context
+from .log import getLogger
+
+log = getLogger(__name__)
 
 # Reference : http://lxr.free-electrons.com/source/arch/x86/include/asm/sigcontext.h?v=2.6.28#L138
-_registers_32 = ["gs",   "fs",  "es",  "ds",   "edi",  "esi", "ebp", "esp", "ebx",
+_registers_i386 = ["gs",   "fs",  "es",  "ds",   "edi",  "esi", "ebp", "esp", "ebx",
```
Comment (zachriggle) [inline pwnlib/srop.py]: registers = {
'i386': ...

File: pwnlib/srop.py [hunk 2]
```diff
@@ -1,26 +1,27 @@
...
         "edx",  "ecx", "eax", "trapno", "err", "eip", "cs",  "eflags",
         "esp_at_signal", "ss",  "fpstate"]
 
 # Reference : https://www.cs.vu.nl/~herbertb/papers/srop_sp14.pdf
-_registers_64 = ["uc_flags", "&uc", "uc_stack.ss_sp", "uc_stack.ss_flags", "uc_stack.ss_size",
+_registers_amd64 = ["uc_flags", "&uc", "uc_stack.ss_sp", "uc_stack.ss_flags", "uc_stack.ss_size",
```
Comment (zachriggle) [inline pwnlib/srop.py]: 'amd64': ...

File: pwnlib/srop.py [hunk 3]
```diff
@@ -1,26 +1,27 @@
...
-        'amd64' : {},
-        'i386'  : {}
-        }
-
-for pos, reg in enumerate(_registers_32):
-    _reg_pos_mapping['i386'][reg] = pos
-
-for pos, reg in enumerate(_registers_64):
-    _reg_pos_mapping['amd64'][reg] = pos
+def get_registers(**kwargs):
```
Comment (zachriggle) [inline pwnlib/srop.py]: This whole function can go away

File: pwnlib/srop.py [hunk 4]
```diff
@@ -68,46 +69,44 @@ class SigreturnFrame(object):
         [0, 0, 0, 0, 0, 0, 0, 0, 6295552, 7, 4096, 125, 0, 0, 0, 115, 0, 0, 123, 0]
     """
 
-    def __init__(self, arch="i386"):
-        self.arch  = arch
+    def __init__(self, **kwargs):
```
Comment (zachriggle) [inline pwnlib/srop.py]: Need to actually use kwargs :)

File: pwnlib/srop.py [hunk 5]
```diff
@@ -68,46 +69,44 @@ class SigreturnFrame(object):
...
-    def _initialize_vals(self):
-        if self.arch == "i386":
-            self._initialize_i386()
-        elif self.arch == "amd64":
-            self._initialize_amd64()
-
-    def _initialize_amd64(self):
-        for i in range(len(_registers_64)):
+    def _initialize_vals(self, **kwargs):
+        values_to_set = { "i386" : [("cs", 0x73), ("ss", 0x7b)],
```
Comment (zachriggle) [inline pwnlib/srop.py]: I would pull out `values_to_set` into the global scope as `defaults`.  For example:

```
defaults = { 'i386': { 'cs': 0x73, 'ss': 0x7b }, 'amd64': ... }
```

File: pwnlib/srop.py [hunk 6]
```diff
@@ -68,46 +69,44 @@ class SigreturnFrame(object):
...
+                          "amd64": [("csgsfs", 0x33)],
+                        }
+        registers = get_registers(**kwargs)
```
Comment (zachriggle) [inline pwnlib/srop.py]: `registers[context.arch]` (after making the change above)

File: pwnlib/srop.py [hunk 7]
```diff
@@ -68,46 +69,44 @@ class SigreturnFrame(object):
...
+        for i in xrange(len(registers)):
             self.frame.append(pack(0x0))
-        self.set_regvalue("csgsfs", 0x33)
 
-    def _initialize_i386(self):
-        for i in range(len(_registers_32)):
-            self.frame.append(pack(0x0))
-        self.set_regvalue("cs", 0x73)
-        self.set_regvalue("ss", 0x7b)
+        with context.local(**kwargs):
```
Comment (zachriggle) [inline pwnlib/srop.py]: `for register, value in defaults[context.arch]`

File: pwnlib/srop.py [hunk 8]
```diff
@@ -68,46 +69,44 @@ class SigreturnFrame(object):
...
-    def set_regvalue(self, reg, val):
+    def set_regvalue(self, reg, val, **kwargs):
         """
         Sets a specific ``reg`` to a ``val``
         """
-        index = _reg_pos_mapping[self.arch][reg]
+        registers = get_registers(**kwargs)
+        index = registers.index(reg)
         value = pack(val)
         self.frame[index] = value
```
Comment (zachriggle) [inline pwnlib/srop.py]: Instead of indexing into self as a frame, it might be better/easier to do as a dictionary, and to perform the packing later.

**Comment by eQu1NoX** [comment]
Hi, I just made the improvements to rop.py and srop.py.

**Comment by eQu1NoX** [comment]
Hi, I just pulled the latest commit from main repo and modified the code to use LocalContext.

**Comment by zachriggle** [comment]
It looks like there are no real changes left in this PR (https://github.com/binjitsu/binjitsu/pull/19/files)

**Pull request closed without merging**